clippy_lints/src/use_self.rs

```diff
@@ -81,78 +69,181 @@ fn span_use_self_lint(cx: &LateContext<'_, '_>, path: &Path<'_>, last_segment: O
+                self.visit_ty(&impl_hir_ty);
+            }
+        }
+    }
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for ImplVisitor<'a, 'tcx> {
     type Map = Map<'tcx>;
 
-    fn visit_ty(&mut self, hir_ty: &'tcx hir::Ty<'_>) {
-        if let TyKind::Path(QPath::Resolved(_, path)) = &hir_ty.kind {
+    fn nested_visit_map(&mut self) -> NestedVisitorMap<Self::Map> {
+        NestedVisitorMap::OnlyBodies(self.cx.tcx.hir())
+    }
+
+    fn visit_ty(&mut self, hir_ty: &'tcx hir::Ty<'tcx>) {
+        if let TyKind::Path(QPath::Resolved(_, path)) = hir_ty.kind {
             match path.res {
                 def::Res::SelfTy(..) => {},
                 _ => {
                     if hir_ty_to_ty(self.cx.tcx, hir_ty) == self.self_ty {
-                        span_use_self_lint(self.cx, path, None);
+                        match self.cx.tcx.hir().find(self.cx.tcx.hir().get_parent_node(hir_ty.hir_id)) {
+                            Some(Node::Expr(Expr {
+                                kind: ExprKind::Path(QPath::TypeRelative(_, last_segment)),
+                                ..
+                            })) =>
+                            // FIXME: this span manipulation should not be necessary
+                            // @flip1995 found an ast lowering issue in
+                            // https://github.com/rust-lang/rust/blob/master/src/librustc_ast_lowering/path.rs#L142-L162
+                            {
+                                span_lint(self.cx, hir_ty.span.with_hi(last_segment.ident.span.lo() - BytePos(2)))
+                            },
+                            _ => span_lint(self.cx, hir_ty.span),
+                        }
                     }
                 },
             }
         }
 
-        walk_ty(self, hir_ty)
-    }
-
-    fn nested_visit_map(&mut self) -> NestedVisitorMap<Self::Map> {
-        NestedVisitorMap::None
+        walk_ty(self, hir_ty);
     }
-}
-
-fn check_trait_method_impl_decl<'a, 'tcx>(
-    cx: &'a LateContext<'a, 'tcx>,
-    impl_item: &ImplItem<'_>,
-    impl_decl: &'tcx FnDecl<'_>,
-    impl_trait_ref: ty::TraitRef<'tcx>,
-) {
-    let trait_method = cx
-        .tcx
-        .associated_items(impl_trait_ref.def_id)
-        .find_by_name_and_kind(cx.tcx, impl_item.ident, ty::AssocKind::Fn, impl_trait_ref.def_id)
-        .expect("impl method matches a trait method");
 
-    let trait_method_sig = cx.tcx.fn_sig(trait_method.def_id);
-    let trait_method_sig = cx.tcx.erase_late_bound_regions(&trait_method_sig);
-
-    let output_hir_ty = if let FnRetTy::Return(ty) = &impl_decl.output {
-        Some(&**ty)
-    } else {
-        None
-    };
+    fn visit_expr(&mut self, expr: &'tcx Expr<'tcx>) {
+        fn expr_ty_matches<'a, 'tcx>(expr: &'tcx Expr<'tcx>, self_ty: Ty<'tcx>, cx: &LateContext<'a, 'tcx>) -> bool {
+            cx.tcx
+                .typeck_tables_of(expr.hir_id.owner.to_def_id())
+                .node_type(expr.hir_id)
```

Comment:
This is responsible for the ICE in `ui/missing_const_for_fn/could_be_const.rs`. Try `expr_ty_opt` here to possibly prevent this.

Reply:
Oh, what I forget to tell you, is that you might call `cx.tcx.has_typeck_tables(def_id)` before trying to get the typeck tables. At least this is how it's done in most of the rest of Clippy.

Reply:
Thank for the `cx.tcx.has_typeck_tables(def_id)` tip. I'll give it a try.
`typeck_tables_of` is documented in the rustc guide. I tried `expr_ty` without success, the tests panicked.

Reply:
If `expr_ty` panics, `node_ty` will also panic, since `expr_ty` calls `node_ty`. You have to use `if let Some(ty) = tcx.typeck_tables_of(def_id).expr_ty_opt(expr)`, but first make sure to check `cx.tcx.has_typeck_tables(def_id)`.

Reply:
I've should have protected all occurrences of type expression queries with `has_typeck_tables`. The problem remains.